cmd/object-handlers.go

```diff
@@ -2804,10 +2870,21 @@ func (api objectAPIHandlers) DeleteObjectHandler(w http.ResponseWriter, r *http.
 	if _, err := globalBucketMetadataSys.GetLifecycleConfig(bucket); err == nil {
 		hasLifecycleConfig = true
 	}
+
+	var goiOpts ObjectOptions
+	os := newObjSweeper(bucket, object).ForDelete(singleDelete(*r))
+	if hasLifecycleConfig {
+		// Mutations of objects on versioning suspended buckets
+		// affect its null version. Through opts below we select
+		// the null version's remote object to delete if
+		// transitioned.
+		goiOpts = os.GetOpts()
```

Comment:
if versioning is suspended on the bucket, null version of object (which may not be at the top of the stack) will be permanently deleted instead of setting a delete marker.

Reply:
As discussed over the call, null version is handled correctly. I have added a comment to objSweeper.ReqVersion to clarify its origin.